File: src/types/sync.ts
import type { Server } from "@/types/ssh";

/**
 * Contrat de synchronisation (Phase 2). Défini dès maintenant pour que le
 * frontend n'ait AUCUNE modification à subir quand la couche réseau (NestJS +
 * JWT + PostgreSQL + Redis) sera branchée : il suffira de fournir une autre
 * implémentation de `SyncProvider`.
 */

/** Un snippet de commande réutilisable (préparé pour la sync). */
export interface Snippet {
  id: string;
  name: string;
  command: string;
  description?: string;
}

/** Instantané synchronisable de l'état local. */
export interface SyncableSnapshot {
  servers: Server[];
  snippets: Snippet[];
  preferences: Record<string, unknown>;
}

export type SyncStatus = "disabled" | "idle" | "syncing" | "error";

/** Fournisseur de synchronisation. */
export interface SyncProvider {
  readonly name: string;
  isEnabled(): boolean;
  status(): Promise<SyncStatus>;
  /** Récupère l'état distant. */
  pull(): Promise<SyncableSnapshot>;
  /** Pousse l'état local. */
  push(snapshot: SyncableSnapshot): Promise<void>;
}

File: src-tauri/src/services/sync.rs
//! Interfaces de synchronisation (Phase 2) — DÉFINITIONS SEULEMENT.
//!
//! Aucune implémentation réseau ici : la couche cloud (NestJS + JWT +
//! PostgreSQL + Redis) sera branchée plus tard en fournissant un type qui
//! implémente [`SyncProvider`], sans modifier le reste du backend.

use crate::errors::AppError;
use crate::models::Server;

/// Instantané synchronisable de l'état local.
#[derive(Debug, Clone)]
pub struct SyncSnapshot {
    pub servers: Vec<Server>,
    // snippets / préférences seront ajoutés avec la Phase 2.
}

/// Fournisseur de synchronisation (contrat Phase 2).
#[allow(dead_code)]
pub trait SyncProvider {
    fn is_enabled(&self) -> bool;
    fn pull(&self) -> Result<SyncSnapshot, AppError>;
    fn push(&self, snapshot: &SyncSnapshot) -> Result<(), AppError>;
}
